gk/co.c

```diff
@@ -0,0 +1,165 @@
+yield_next(struct gk_co *this_co)
+{
+	aco_yield_to(get_next_co(this_co)->aco);
+}
+
+static inline void
+reset_leftover(struct gk_co_leftover *leftover)
+{
+	leftover->fe = NULL;
+	leftover->fib = NULL;
+}
+
+/*
+ * CAUTION: callers must be aware that when this function returns,
+ * @this_co->work might have changed!
+ */
+static struct gk_co_task *
+get_the_very_first_task(struct gk_co *this_co, struct gk_co ***p_this_co_bucket)
+{
+	struct gk_co_work *work = this_co->work;
+	struct gk_co_task *first_task = NULL;
+
+	do {
+		struct gk_co_task *task, *temp_task;
+
+		while (unlikely(list_empty(&work->tasks))) {
+			/* There is no work to do. */
+
+			struct gk_co *next_co = get_next_co(this_co);
+
+			/* Make @this_co idle. */
+			list_del(&this_co->cos_list);
+			list_add_tail(&this_co->cos_list, &work->idle_cos);
+
+			if (likely(this_co != next_co)) {
+				/*
+				 * @this_co is NOT the last working coroutine.
+				 * Yield to the next coroutine.
+				 */
+				aco_yield_to(next_co->aco);
+			} else {
+				/*
+				 * No more work and no more working coroutines.
+				 * Return to the main coroutine.
+				 */
+				aco_yield();
+			}
+		}
+
+		/* Find a task to assign to @this_co. */
+		list_for_each_entry_safe(task, temp_task, &work->tasks,
+				task_list) {
+			struct gk_co *task_owner_co;
+			uint16_t task_masked_hash;
+
+			list_del(&task->task_list);
+			task_masked_hash = task->task_hash & work->hash_mask;
+			task_owner_co = work->hash_to_assigned_co
+				[task_masked_hash];
+			if (task_owner_co == NULL) {
+				if (first_task == NULL) {
+					/* A task to call our own. */
+					first_task = task;
+					/*
+					 * Take over the following tasks that
+					 * share the same masked hash.
+					 */
+					*p_this_co_bucket =
+						&work->hash_to_assigned_co
+						[task_masked_hash];
+					**p_this_co_bucket = this_co;
+				} else {
+					break;
+				}
+			} else {
+				/* Assign this task to its owner. */
+				list_add_tail(&task->task_list,
+					&task_owner_co->task_queue);
+			}
+		}
+	} while (first_task == NULL);
+
+	return first_task;
+}
+
+/*
+ * CAUTION: callers must be aware that when this function returns,
+ * @this_co->work might have changed!
+ */
+static struct gk_co_task *
+next_task(struct gk_co *this_co, struct gk_co ***p_this_co_bucket,
+	struct gk_co_leftover *leftover)
+{
+	if (!list_empty(&this_co->task_queue)) {
+		/*
+		 * @this_co already has assigned tasks.
+		 * Return the first assigned task.
+		 */
+		struct gk_co_task *task = list_first_entry(&this_co->task_queue,
+			struct gk_co_task, task_list);
+		list_del(&task->task_list);
+		return task;
+	}
+
+	/* There is no more tasks assigned to @this_co. */
+	**p_this_co_bucket = NULL;	/* NULL bucket entry.	*/
+	*p_this_co_bucket = NULL;	/* NULL bucket tracker.	*/
+	reset_leftover(leftover);
+
+	return get_the_very_first_task(this_co, p_this_co_bucket);
+}
+
+void
+gk_co_main(void)
+{
+	struct gk_co *this_co = aco_get_arg();
+	struct gk_co_leftover leftover = {};
```

Comment:
Correct me if I am wrong: it seems that we are using only one ``leftover`` per coroutine. If so, what if two packets from two different flows are mapped to the same coroutine and these two packets may have different flow entries as well as fib entries for processing it?

Reply:
You are right, there is only one `leftover` per coroutine.

The case you talking about is a false flow association. This false association is unlikely because `work->hash_to_assigned_co` is "large", but it can happen. I chose the current algorithm because all other solutions that I explored had a **much** higher overhead.

When a false association does happen, it's the responsibility of the coroutine to ignore the `leftover`. See an example in `gk_co_process_front_pkt_final()`.

Reply:
I see, indeed ``gk_co_process_front_pkt_final()`` handles the above case correctly.

I wonder is it possible to keep the ``leftover`` as part of the ``struct ipacket``? Since all tasks processing the same packet should share these information, while the packet itself has all the property of carrying these information.

Reply:
`leftover` is to be shared among all packets of the same flow, so we can shortcut the processing of the packets in the same flow. If we limit `leftover` to a single packet by making it part of `struct ipacket`, our goal is lost.

Reply:
Notice that `leftover` is not required to make the code work, it's an optimization. In our evaluations, we use a uniform distribution over the source addresses of the packets, so `leftover` is likely not being used at all. But I expect that, in production, lots of batches of packets will have at least two packets of the same flow. Each time `leftover` is reused is a big saving because it saves at least one lookup to the flow table. When the flow table is full, reusing the `leftover` saves one lookup to the flow table **and** a lookup to the LPM table because the temporary flow entry is reused!